XXX-asyncwrap-api.md

```diff
@@ -0,0 +1,309 @@
+| Title  | AsyncWrap API |
+|--------|---------------|
+| Author | @trevnorris   |
+| Status | DRAFT         |
+| Date   | 2016-04-05    |
+
+## Description
+
+Since its initial introduction with the `AsyncListener` API, `AsyncWrap` has
+slowly evolved to ensure a generalized API that would serve as a solid base for
+module authors who wished to add events to the life cycle of the event loop.
+The API presented below aims to accomplish this.
+
+## API
+
+### Overview
+
+```js
+// Standard way of requiring a module. Snake case follows core module
+// convention.
+const async_wrap = require('async_wrap');
+
+// List of asynchronous providers (e.g. CRYPTO or TCPWRAP). Can be used for
+// trace filtering of events.
+const providers = async_wrap.Providers;
+
+// init() is called during object construction. The object handle being
```

Comment:
> For example, doesn't `init()` really mean `enqueueAnAsyncCallback()`. In the case of nextTick callbacks & promises, it doesn't correspond to an object being constructed, right?

Yes it does. It always does. It's always constructed at the time of call. You're confusing `init()` with something like when a callback is passed to the instances event handler, e.g. `on('data'` (which is the JS abstraction of the handle and has nothing to do with how the handle operates). The callback that will be called by `MakeCallback` is always defined at construction time. That will then trigger any callbacks from the user that have been abstracted via the JS API.